cpu/esp32/periph/i2c_hw.c

```diff
@@ -182,38 +177,64 @@ void i2c_init(i2c_t dev)
 
     /* determine the half period of clock in APB clock cycles */
     uint32_t half_period = 0;
-
-    switch (_i2c_bus[dev].speed) {
-        case I2C_SPEED_LOW:
-            /* 10 kbps (period 100 us) */
-            half_period = (I2C_CLK_FREQ / 10000) >> 1;
-            break;
-
-        case I2C_SPEED_NORMAL:
-            /* 100 kbps (period 10 us) */
-            half_period = (I2C_CLK_FREQ / 100000) >> 1;
-            half_period = half_period * 95 / 100; /* correction factor */
-            break;
-
-        case I2C_SPEED_FAST:
-            /* 400 kbps (period 2.5 us) */
-            half_period = (I2C_CLK_FREQ / 400000) >> 1;
-            half_period = half_period * 82 / 100; /* correction factor */
-            break;
-
-        case I2C_SPEED_FAST_PLUS:
-            /* 1 Mbps (period 1 us) not working */
-            half_period = (I2C_CLK_FREQ / 1000000) >> 1;
-            break;
-
-        case I2C_SPEED_HIGH:
-            /* 3.4 Mbps (period 0.3 us) not working */
-            half_period = (I2C_CLK_FREQ / 3400000) >> 1;
-            break;
-
-        default:
-            LOG_TAG_ERROR("i2c", "Invalid speed value in %s\n", __func__);
-            return;
+    uint32_t apb_clk = rtc_clk_apb_freq_get();
+
+    if (apb_clk == MHZ(2)) {
+        /* CPU clock frequency of 2 MHz requires special handling */
+        switch (_i2c_bus[dev].speed) {
+            case I2C_SPEED_LOW:
+                /* 10 kbps (period 100 us) */
+                half_period = 95;
+                break;
+
+            case I2C_SPEED_NORMAL:
+                /* 100 kbps (period 10 us) */
+                /* NOTE: Correct value for half_period would be 6 to produce a
+                 *       100 kHz clock. However, a value of at least 18 is
+                 *       necessary to work correctly which corresponds to a
+                 *       I2C clock speed of 30 kHz.
+                 */
+                half_period = 18;
+                break;
+
+            default:
+                LOG_TAG_ERROR("i2c", "I2C clock speed not supported in "
+                                     "hardware with CPU clock 2 MHz, use the "
+                                     "software implementation instead\n");
+                assert(0);
+        }
+    }
+    else {
+        switch (_i2c_bus[dev].speed) {
+            case I2C_SPEED_LOW:
+                /* 10 kbps (period 100 us) */
+                half_period = (apb_clk / 10000) >> 1;
+                break;
+
+            case I2C_SPEED_NORMAL:
+                /* 100 kbps (period 10 us) */
+                half_period = (apb_clk / 100000) >> 1;
```

Comment:
Are the corrections not needed anymore?

Reply:
The former correction factors were determined by measuring the resulting clocks without a device connected to the bus. 

However, when I tested the changes for low CPU clock frequencies, I measured the resulting clocks without and with a device connected to the bus. I found that the clocks were different and seemed to depend on the bus capacity. It is clear that the rise time of I2C signals depends on the bus capacities, since the outputs are pulled up open-drain outputs.

So it seems that
```
    _i2c_hw[dev].regs->scl_low_period.period = half_period;
    _i2c_hw[dev].regs->scl_high_period.period = half_period;
```
are not absolute times, but times that start when the corresponding level is reached. If this is the case, the higher the bus capacity, the longer the period would be. This means that we would not be able to control the clock speed precisely.

Because of this and since the I2C implementation in ESP-IDF don't use any correction factor, I removed them.

Reply:
Sounds good. Can you update the commit message with this?

Ready to squash.